# Prediction Challenge! of the AIBBC 2025 Bioinformatics and Genomics 

## Load library and data

```{r}
library(tidyverse)
library(tidymodels)
library(pROC)
```

```{r}
df_train <- read_rds(file = "~/Documents/AIBBC_2025/Challenge/Data/tcga_train.rds") 
df_test <- read_rds(file = "~/Documents/AIBBC_2025/Challenge/Data/tcga_test.rds") 
```

## Quick view of all data

```{r}
df_train[1:10, 1:10]
```

```{r}
df_test[1:10, 1:10]
```

You will see that `df_test$response` and `df_test$tissue` column is `NA`.

The aim of this challenge is to predict:

1)  "**response",** whether the sample comes from Tumor or Normal tissue, and

2)  "**tissue**", of which the sample comes from.

### We will provide you a backbone for this challenge!

## Better overview using PCA plot

```{r}
df_all <- rbind(df_train, df_test)
df_exp_only <- subset(df_all, select = -c(response, tissue))
```

```{r}
df_pca_only <- prcomp(df_exp_only, scale=TRUE)
```

```{r}
df_pca_all <- augment(df_pca_only, subset(df_all, select = c(response, tissue)))
df_pca_all <- column_to_rownames(df_pca_all, var = ".rownames")
```

### Plot the PCA

```{r}
df_pca_all %>% 
  ggplot(aes(x=.fittedPC1, y=.fittedPC2)) +
    geom_point(aes(colour = tissue)) +
    theme_classic()
```

```{r}
df_pca_all %>% 
  ggplot(aes(x=.fittedPC1, y=.fittedPC2)) +
    geom_point(aes(colour = response)) +
    theme_classic()
```

```{r}
df_pca_all %>% 
  filter(!is.na(response)) %>% 
  ggplot(aes(x=.fittedPC1, y=.fittedPC2)) +
    geom_point(aes(colour = tissue)) +
    theme_classic()
```

```{r}
df_pca_all %>% 
  filter(!is.na(response)) %>% 
  ggplot(aes(x=.fittedPC1, y=.fittedPC2)) +
    geom_point(aes(colour = response)) +
    theme_classic()
```

## Build the model!

### In this example, we will use simple logistic regression. However, you are welcome to choose your own model!

```{r}

# Always make the same split
set.seed(0)

# We use 80% for training and 20% for validation
trainfold <- df_train %>% sample_frac(size=0.80)
valfold  <- setdiff(df_train, trainfold)

# We train our glm (in this example we use simple logistic regression)
# We also will only use some genes for example
fit <- logistic_reg(mode = "classification") %>%
  set_engine("glm") %>%
  fit(response ~ TRIM48 * SPRR4 * OR8K1, data = trainfold)
```

```{r}
summary(fit)
```

```{r}
# We predict on the validation fold
predicted_class <- predict(fit, new_data = valfold, type = "class")

# Combine predictions with the observed values
pred_df <- cbind(valfold, predicted_class)

# Compute accuracy directly using yardstick
accuracy(pred_df, truth = response, estimate = .pred_class)
```

```{r}
# Or compute the full confusion matrix
conf_mat(pred_df, truth = response, estimate = .pred_class)
```

```{r}

# Encode Tumor = 1, Normal = 0 in a compact and consistent way
pred_encoded <- pred_df %>%
  mutate(
    pred_numeric = as.integer(.pred_class == "Tumor"),
    truth_numeric = as.integer(response == "Tumor")
  )

roc_result <- roc(pred_encoded$truth_numeric, pred_encoded$pred_numeric)

# Print ROC result
print(roc_result)
```

```{r}

# Optional: plot ROC curve
plot(roc_result, col = "blue", main = "ROC Curve for Logistic Regression")
```

## Make your actual prediction on test set: the actual unknown data!

Previously, we fit the model to all of our known data.

Now, we want to predict on the unknown data.

The predictions must have the following column and the row order must be the same as the original!

-   prediction (the predicted class, either "Normal" or "Tumor")

```{r}

my_prediction <- predict(fit, new_data = df_test, type = "class")

submission <- tibble(
  id = rownames(df_test),                 # store IDs here
  prediction = my_prediction$.pred_class
)

head(submission)
```

## Submitting your answer

The following code will give us

-   your chosen team name
-   the name of the people on the team
-   your estimated accuracy
-   your predictions

Please edit the values below .

The filename of the output will be automated challenge_response.TEAMNAME.rds

Please - do not use space or funny letters in your team name.

```{r}

team_name        <- "team_aarhus" 
team_people      <- c("Jakob", "Farhad") 
team_error       <- 0.1
team_predictions <- submission
try <- 1

#
# Always run this code
# If it fails you have done something wrong!
#
# Extract the columns needed
team_predictions <- team_predictions %>% select(prediction)

# Save all the stuff in one object
write_rds(x = list(team_name, team_people, team_error, team_predictions, try), 
          file = paste("challenge_response.", team_name, ".rds", sep=""))
```

## Don't forget to submit your rds prediction file into Google Drive!

# Good luck!
